lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp

```diff
@@ -398,28 +394,28 @@ ScanOpConversion::getDelinearizedIds(ConversionPatternRewriter &rewriter,
 
   auto threadsPerWarp = triton::gpu::getThreadsPerWarp(srcEncoding);
   auto warpsPerCTA = triton::gpu::getWarpsPerCTA(srcEncoding);
-  auto threadOrder = triton::gpu::getThreadOrder(srcEncoding);
-  auto warpOrder = triton::gpu::getWarpOrder(srcEncoding);
-  SmallVector<Value> multiDimLaneId =
-      delinearize(rewriter, loc, laneId, threadsPerWarp, threadOrder);
-  SmallVector<Value> multiDimWarpId =
-      delinearize(rewriter, loc, warpId, warpsPerCTA, warpOrder);
+  auto [multiDimLaneId, isRepresentativeLane] =
+      getMultiDimLaneId(rewriter, helper, laneId);
+  auto [multiDimWarpId, isRepresentativeWarp] =
+      getMultiDimWarpId(rewriter, helper, warpId);
 
   Value laneIdAxis = multiDimLaneId[axis];
   Value warpIdAxis = multiDimWarpId[axis];
 
   multiDimLaneId[axis] = b.i32_val(0);
   threadsPerWarp[axis] = 1;
-  Value laneIdParallel =
-      linearize(rewriter, loc, multiDimLaneId, threadsPerWarp, threadOrder);
+  Value laneIdParallel = linearize(rewriter, loc, multiDimLaneId,
+                                   threadsPerWarp, helper.getOrder());
   multiDimWarpId[axis] = b.i32_val(0);
   warpsPerCTA[axis] = 1;
   Value warpIdParallel =
-      linearize(rewriter, loc, multiDimWarpId, warpsPerCTA, warpOrder);
+      linearize(rewriter, loc, multiDimWarpId, warpsPerCTA, helper.getOrder());
   Value flatIdParallel = b.add(
       laneIdParallel,
       b.mul(warpIdParallel, b.i32_val(helper.getNonAxisNumThreadsPerWarp())));
-  return std::make_tuple(laneIdAxis, warpIdAxis, flatIdParallel);
+  auto isRepresentative = b.and_(isRepresentativeLane, isRepresentativeWarp);
```

Comment:
what does `isRepresentative` means? Why do we need to change the codegen if we are not supporting linear layout here? Are the APIs not returning the same values anymore?

Reply:
This handles the case when different threads hold the same value (broadcasting). `isRepresentative` is just true for the first thread that holds the value. This prevents from doing different STS that may not be necessary.

I added this while debugging certain annoying issue, but in retrospect, it may not be 100% necessary for this PR. That being said, I think it's a good thing to have.

Reply:
Is this similar to what Jeff had fixed then? It's a good thing to have because you think it solves some latent bugs?

Reply:
Yes. I do not think that Jeff's fix was fully correct, but I very much think this one is.

Reply:
oh I see, do we know what case it fixes? Feels a bit weird to add code fixing something we don't know how to test. It is very likely to break next time we change the code

Reply:
I had to move to these functions to be able to move away from some legacy methods, in particular `getThreadsPerWarp` and `getWarpsPerCTA`. The latent bug is that I think that the fix that Jeff implemented for threads (in lines 483-491 in the previous file) was still needed for warps. All these points become moot with the new implementation, as the `delinearize` function I added to fix `ReduceOp(LinearEncoding)` handles all this in full generality.

Reply:
Now that I think about it, we should be able to reduce the amount of shmem reduce and scan use, as now delinearize computes indices between 0..prod(getThreadsPerWarp)-1 which may be less than 32 if there is broadcasting! Before we were using indices from 0..31 with perhaps holes in them if there was broadcasting.
I'll send a PR fixing this at some point.